presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+        }
+        switch (oldTableAction.getType()) {
+            case DROP:
+                throw new TableNotFoundException(schemaTableName);
+            case ADD:
+            case ALTER:
+            case INSERT_EXISTING:
+                throw new UnsupportedOperationException("dropping a table added/modified in the same transaction is not supported");
+            default:
+                throw new IllegalStateException("Unknown action type");
+        }
+    }
+
+    public synchronized void replaceView(String databaseName, String tableName, Table table, PrincipalPrivilegeSet principalPrivilegeSet)
+    {
+        setExclusive((delegate, hdfsEnvironment) -> delegate.replaceTable(databaseName, tableName, table, principalPrivilegeSet));
+    }
+
+    public synchronized void renameTable(String databaseName, String tableName, String newDatabaseName, String newTableName)
+    {
+        setExclusive((delegate, hdfsEnvironment) -> delegate.renameTable(databaseName, tableName, newDatabaseName, newTableName));
+    }
+
+    public synchronized void addColumn(String databaseName, String tableName, String columnName, HiveType columnType, String columnComment)
+    {
+        setExclusive((delegate, hdfsEnvironment) -> delegate.addColumn(databaseName, tableName, columnName, columnType, columnComment));
+    }
+
+    public synchronized void renameColumn(String databaseName, String tableName, String oldColumnName, String newColumnName)
+    {
+        setExclusive((delegate, hdfsEnvironment) -> delegate.renameColumn(databaseName, tableName, oldColumnName, newColumnName));
+    }
+
+    public synchronized void finishInsertIntoExistingTable(ConnectorSession session, String databaseName, String tableName, Path currentLocation, List<String> fileNames)
+    {
+        setShared();
+        SchemaTableName schemaTableName = new SchemaTableName(databaseName, tableName);
+        Action<TableAndMore> oldTableAction = tableActions.get(schemaTableName);
+        if (oldTableAction == null) {
+            Optional<Table> table = delegate.getTable(databaseName, tableName);
+            if (!table.isPresent()) {
+                throw new TableNotFoundException(schemaTableName);
+            }
+            tableActions.put(
+                    schemaTableName,
+                    new Action<>(
+                            ActionType.INSERT_EXISTING,
+                            new TableAndMore(table.get(), Optional.empty(), Optional.of(currentLocation), Optional.of(fileNames)), session.getUser(), session.getQueryId()));
+            return;
+        }
+
+        switch (oldTableAction.getType()) {
+            case DROP:
+                throw new TableNotFoundException(schemaTableName);
+            case ADD:
+            case ALTER:
+            case INSERT_EXISTING:
+                throw new UnsupportedOperationException("Inserting into an unpartitioned table that were added, altered, or inserted into in the same transaction is not supported");
```

Comment:
Because data can never be inserted into a partitioned table. Data can only be inserted into partitions and unpartitioned tables.